modules/transactionpool/transactionpool.go

```diff
@@ -9,86 +9,53 @@ import (
-	siafundOutputs map[types.SiafundOutputID]types.SiafundOutput
-
-	// The reference set contains any objects that are not in the unconfirmed
-	// set, but may still need to be referenced when creating diffs or
-	// reverting unconfirmed transactions (due to conflicts).
-	referenceSiacoinOutputs        map[types.SiacoinOutputID]types.SiacoinOutput
-	referenceFileContracts         map[types.FileContractID]types.FileContract
-	referenceFileContractRevisions map[crypto.Hash]types.FileContract
-	referenceSiafundOutputs        map[types.SiafundOutputID]types.SiafundOutput
-
-	// The entire history of the transaction pool is kept. Each element
-	// represents an atomic change to the transaction pool. When a new
-	// subscriber joins the transaction pool, they can be sent the entire
-	// history and catch up properly, and they can take a long time to catch
-	// up. To prevent deadlocks in the transaction pool, subscribers are
-	// updated in a separate thread which does not guarantee that a subscriber
-	// is always fully synchronized to the transaction pool.
-	consensusChangeIndex    int   // Increments any time a consensus update is made.
-	consensusChanges        []int // The index of the consensus change from the consensus set. -1 means there was no change from the consensus set.
-	unconfirmedTransactions [][]types.Transaction
-	unconfirmedSiacoinDiffs [][]modules.SiacoinOutputDiff
-	subscribers             []chan struct{}
-
-	mu *sync.RWMutex
-}
+type (
+	// ObjectIDs are the IDs of objects such as siacoin outputs and file
+	// contracts, and are used to see if there are conflicts or overlaps within
+	// the transaction pool. A TransactionSetID is the hash of a transaction
+	// set.
+	ObjectID         crypto.Hash
+	TransactionSetID crypto.Hash
+
+	// The TransactionPool tracks incoming transactions, accepting them or
+	// rejecting them based on internal criteria such as fees and unconfirmed
+	// double spends.
+	TransactionPool struct {
+		// Depedencies of the transaction pool.
+		consensusSet modules.ConsensusSet
+		gateway      modules.Gateway
+
+		// To prevent double spends in the unconfirmed transaction set, the
+		// transaction pool keeps a list of all objects that have either been
+		// created or consumed by the current unconfirmed transaction pool. All
+		// transactions with overlaps are rejected. This model is
+		// over-aggressive - one transaction set may create an object that
+		// another transaction set spends. This is done to minimize the
+		// computation and memory load on the transaction pool. Dependent
+		// transactions should be lumped into a single transaction set.
+		//
+		// transactionSetDiffs map form a transaction set id to the set of
+		// diffs that resulted from the transaction set.
+		knownObjects        map[ObjectID]struct{}
+		transactionSets     map[TransactionSetID][]types.Transaction
+		transactionSetDiffs map[TransactionSetID]modules.ConsensusChange
+		databaseSize        int
```

Comment:
oh. I waffled on the name for a while before settling on transactionList

so the right name I think would be listSize or transactionListSize